### ISSUE TITLE
Strange formatting of `@`, `|`, and parens patterns

### ISSUE THREAD

**Issue by Centril** [rust-lang/rustfmt#4110]
In a rust-lang/rust PR, I saw some pretty strange formatting of pattern matching with `@`, `|`, and parens patterns: https://github.com/rust-lang/rust/pull/70906/files#r405165086

The diff was:

```diff
-                    category: category @ ConstraintCategory::Return,
+                    category:
+                        category
+                        @
+                        (ConstraintCategory::Return
+                        | ConstraintCategory::CallArgument
+                        | ConstraintCategory::OpaqueType),
```

which seems odd because it is putting `@` on its own line and doing `(` on the same line as a pattern and then `)` also on the same line as a pattern.

I would have expected:

```diff
-                    category: category @ ConstraintCategory::Return,
+                    category: category @ (
+                        ConstraintCategory::Return
+                        | ConstraintCategory::CallArgument
+                        | ConstraintCategory::OpaqueType
+                    ),
```

**Comment by calebcartwright** [comment]
Not a duplicate but similar report in #4031. Fixes will likely be similar (if not identical)

**Comment by topecongiro** [comment]
Did this pattern exist before, or is it something new that has been introduced recently? Either way, we should fix this.

**Comment by calebcartwright** [comment]
I believe it's a new-ish change, but not sure precisely when it was introduced

**Comment by Centril** [comment]
```rust
category: category @ Record { ... }
```
has existed since Rust 1.0 I think, so that isn't new, but `p0 | ... | pn` nested within another pattern is new and currently an unstable feature.

**Comment by calebcartwright** [comment]
Oops, I misunderstood the question anyway. I don't know definitively but I believe that the poor binding formatting described in #4031 was a change rustfmt started doing relatively recently.

Not sure what, if any, impact that may have on the formatting here

**Comment by qm3ster** [comment]
Happens equally as root for a match arm:
```rust
match head.packet_type()? {
    p
    @
    (PacketType::Connect
    | PacketType::ConnAck
    | PacketType::SubAck
    | PacketType::UnsubAck) => return Err(Error::WrongPacket(p)),
```
Manually, I would probably format it as:
```rust
match head.packet_type()? {
    p @ (
        PacketType::Connect
        | PacketType::ConnAck
        | PacketType::SubAck
        | PacketType::UnsubAck
    ) => return Err(Error::WrongPacket(p)),
 ```
 Interestingly enough, we still end up with N+2 lines. But moving the output expression to the left is definitely a huge win.

**Comment by evanjs** [comment]
"Fixing" this by locally by increasing `max_width` for now (not really a fix...) but this is definitely something that would make rustfmt _much_ less annoying when using nested or patterns.

I wonder if this would be more worthwhile to investigate now that https://github.com/rust-lang/rust/issues/54883 is stabilized (?), etc.

**Comment by ytmimi** [comment]
This was fixed in #4978, and a test case was added for this issue
